crates/block-producer/src/batch_builder/mod.rs

```diff
@@ -67,280 +66,246 @@ impl BatchBuilder {
     /// available to batch.
-    pub async fn run(self, mempool: SharedMempool, store: StoreClient) {
+    pub async fn run(mut self, mempool: SharedMempool) {
         assert!(
             self.failure_rate < 1.0 && self.failure_rate.is_sign_positive(),
             "Failure rate must be a percentage"
         );
 
         let mut interval = tokio::time::interval(self.batch_interval);
-        interval.set_missed_tick_behavior(time::MissedTickBehavior::Delay);
-
-        let mut worker_pool = WorkerPool::new(
-            self.workers,
-            self.simulated_proof_time,
-            self.failure_rate,
-            store,
-            self.batch_prover,
-        );
+        // We set the inverval's missed tick behaviour to burst. This means we'll catch up missed
+        // batches as fast as possible. In other words, we try our best to keep the desired batch
+        // interval on average. The other options would result in at least one skipped batch.
+        interval.set_missed_tick_behavior(time::MissedTickBehavior::Burst);
 
         loop {
-            tokio::select! {
-                _ = interval.tick() => {
-                    if !worker_pool.has_capacity() {
-                        tracing::info!("All batch workers occupied.");
-                        continue;
-                    }
-
-                    // Transactions available?
-                    let Some((batch_id, transactions)) =
-                        mempool.lock().await.select_batch()
-                    else {
-                        tracing::info!("No transactions available for batch.");
-                        continue;
-                    };
-
-                    worker_pool.spawn(batch_id, transactions).expect("Worker capacity was checked");
-                },
-                result = worker_pool.join_next() => {
-                    let mut mempool = mempool.lock().await;
-                    match result {
-                        Err((batch_id, err)) => {
-                            tracing::warn!(%batch_id, %err, "Batch job failed.");
-                            mempool.batch_failed(batch_id);
-                        },
-                        Ok(batch) => {
-                            mempool.batch_proved(batch);
-                        }
-                    }
-                }
-            }
+            interval.tick().await;
+            self.build_batch(mempool.clone()).await;
         }
     }
-}
 
-// BATCH WORKER
-// ================================================================================================
+    #[instrument(parent = None, target = COMPONENT, name = "batch_builder.build_batch", skip_all)]
+    async fn build_batch(&mut self, mempool: SharedMempool) {
+        Span::current().set_attribute("workers.count", self.worker_pool.len());
 
-type BatchResult = Result<ProvenBatch, (BatchId, BuildBatchError)>;
+        self.wait_for_available_worker().await;
```

Comment:
Could we add a brief comment (either here or as doc comments for the function) explaining how this pattern works? My understanding is that we have a set of dummy tasks in `self.worker_pool`. Here, we check if any of these tasks is "free", and the on lines 100 - 101 below we use the free task run the new batch job?

I'm assuming `wait_for_available_worker()` removes a task from the `JoinSet` while `worker_pool.spawn()` adds a new task to it. Is this correct?

Reply:
Correct. I'll add a comment.

A more conventional pool impl might do something like:

```rust
if pool.workers.len() < pool.max_workers() {
    pool.spawn(...);
} else {
    pool.wait_for_job_completion().await;
    pool.spawn(...);
}
```

So I figured instead of branching we instead just do the `else` branch always. By initializing the pool with `max` "jobs" which are immediately `future::ready()`.